File: src/main/java/com/wikia/tibia/objects/Achievement.java
package com.wikia.tibia.objects;

import com.wikia.tibia.enums.Status;
import com.wikia.tibia.enums.YesNo;
import com.wikia.tibia.utils.ObjectUtils;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

@Getter
@Setter
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class Achievement extends WikiObject {

    private static final Logger LOG = LoggerFactory.getLogger(Achievement.class);

    private Integer grade;
    private String description;
    private String spoiler;
    private YesNo premium;
    private Integer points;
    private YesNo secret;
    private Integer coincideswith;
    private Integer achievementid;
    private String relatedpages;

    @Builder
    public Achievement(String name, String actualname, String implemented, String history, Status status, Integer grade,
                       String description, String spoiler, YesNo premium, Integer points, YesNo secret, Integer coincideswith,
                       Integer achievementid, String relatedpages) {
        super(name, null, actualname, null, implemented, null, history, status);
        this.grade = grade;
        this.description = description;
        this.spoiler = spoiler;
        this.premium = premium;
        this.points = points;
        this.secret = secret;
        this.coincideswith = coincideswith;
        this.achievementid = achievementid;
        this.relatedpages = relatedpages;
    }

    @Override
    public void setDefaultValues() {

        if (ObjectUtils.isEmpty(getImplemented())) {
            setImplemented("?");
        }

        if (ObjectUtils.isEmpty(grade)) {
            LOG.warn("Achievement '{}' has no grade set", getName());
            grade = -1;
        }

        if (ObjectUtils.isEmpty(premium)) {
            premium = YesNo.UNKNOWN;
        }

        if (ObjectUtils.isEmpty(points)) {
            LOG.warn("Achievement '{}' has no points set", getName());
            points = -1;
        }

        if (ObjectUtils.isEmpty(secret)) {
            secret = YesNo.UNKNOWN;
        }

        if (ObjectUtils.isEmpty(achievementid)) {
            LOG.warn("Achievement '{}' has no achievementId set", getName());
            achievementid = -1;
        }
    }
}

File: src/main/kotlin/com/wikia/tibia/objects/Achievement.kt
package com.wikia.tibia.objects

import com.wikia.tibia.enums.Status
import com.wikia.tibia.enums.YesNo
import com.wikia.tibia.utils.ObjectUtils.isEmpty
import org.slf4j.LoggerFactory

data class Achievement(
        val name: String,
        val actualname: String,
        val implemented: String?,
        val history: String,
        val status: Status,
        val grade: Int,
        val description: String,
        val spoiler: String,
        val premium: YesNo?,
        val points: Int?,
        val secret: YesNo?,
        val coincideswith: Int,
        val achievementid: Int?,
        val relatedpages: String
) : WikiObject() {

    override fun setDefaultValues() {
        if (isEmpty(implemented)) {
            this.copy(implemented = "?")
        }
        if (isEmpty(grade)) {
            logger.warn("Achievement '{}' has no grade set", name)
            this.copy(grade = -1)
        }
        if (isEmpty(premium)) {
            this.copy(premium = YesNo.UNKNOWN)
        }
        if (isEmpty(points)) {
            logger.warn("Achievement '{}' has no points set", name)
            this.copy(points = -1)
        }
        if (isEmpty(secret)) {
            this.copy(secret = YesNo.UNKNOWN)
        }
        if (isEmpty(achievementid)) {
            logger.warn("Achievement '{}' has no achievementId set", name)
            this.copy(achievementid = -1)
        }
    }

    companion object {
        private val logger = LoggerFactory.getLogger(Achievement::class.java)
    }
}